File: Difficulty/Medium/[47] Permutations II/solution.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        int size = nums.size();
        permutation(res, 0, size - 1, nums);
        return res;
    }
    
    void permutation(vector<vector<int>>& res, int begin, int len, vector<int>& nums){
        if (begin == len){
            res.push_back(nums);
            return;
        }
        
        unordered_set<int> record;
        for (int i = begin; i <= len; ++i){
            if (record.find(nums[i]) != record.end())
                continue;
            record.insert(nums[i]);
            swap(nums[begin], nums[i]);
            permutation(res, begin + 1, len, nums);
            swap(nums[begin], nums[i]);
        }
    }
};

File: Difficulty/Medium/[47] Permutations II/solution.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = []
        size = len(nums)
        self.permutation(res, 0, size, nums)
        return res
    
    def permutation(self, res, begin, size, nums):
        if begin == size - 1:
            res.append(nums[:])
            return
        record = set()
        for i in range(begin, size):
            if nums[i] in record:
                continue
            record.add(nums[i])
            nums[begin], nums[i] = nums[i], nums[begin]
            self.permutation(res, begin + 1, size, nums)
            nums[begin], nums[i] = nums[i], nums[begin]